lib/media/track/trackpublication.js

```diff
@@ -0,0 +1,91 @@
+'use strict';
+
+const { EventEmitter } = require('events');
+const { buildLogLevels } = require('../../util');
+const { DEFAULT_LOG_LEVEL } = require('../../util/constants');
+const Log = require('../../util/log');
+let nInstances = 0;
+
+/**
+ * A {@link TrackPublication} represents a {@link Track} that
+ * has been published to a {@link Room}.
+ * @property {string} trackName - the published {@link Track}'s name
+ * @property {Track.SID} trackSid - SID assigned to the published {@link Track}
+ * @emits TrackPublication#trackDisabled
+ * @emits TrackPublication#trackEnabled
+ */
+class TrackPublication extends EventEmitter {
+  /**
+   * Construct a {@link TrackPublication}.
+   * @param {boolean} isTrackEnabled - whether the {@link Track} is enabled
+   * @param {string} trackName - the published {@link Track}'s name
+   * @param {Track.SID} trackSid - SID assigned to the {@link Track}
+   * @param {TrackPublicationOptions} options - {@link TrackPublication} options
+   */
+  constructor(isTrackEnabled, trackName, trackSid, options) {
+    super();
+
+    options = Object.assign({
+      logLevel: DEFAULT_LOG_LEVEL
+    }, options);
+
+    const logLevels = buildLogLevels(options.logLevel);
+    let _isTrackEnabled = isTrackEnabled;
+
+    Object.defineProperties(this, {
+      _instanceId: {
+        value: nInstances++
+      },
+      _isTrackEnabled: {
+        get() {
+          return _isTrackEnabled;
+        },
+        set(isTrackEnabled) {
+          if (_isTrackEnabled !== isTrackEnabled) {
+            _isTrackEnabled = isTrackEnabled;
+            this.emit(_isTrackEnabled ? 'trackEnabled' : 'trackDisabled');
+          }
+        }
```

Comment:
I just had a realization. Don't we want the property that

```js
publication.track.on('enabled', () => {
  assert.equal(publication.isTrackEnabled, track.isEnabled);
});

publication.track.on('disabled', () => {
  assert.equal(publication.isTrackEnabled, track.isEnabled);
})
```

Will this design help or hinder fulfilling this property?

Reply:
@markandrus If we want to do this, then the `Track` has to have a set of `.publications` which are publishing the it, and then it has to update each publication's `.isTrackEnabled` whenever its .isEnabled changes.